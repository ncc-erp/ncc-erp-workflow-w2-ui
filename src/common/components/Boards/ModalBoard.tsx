import { useEffect, useMemo, useRef } from 'react';

import {
  AlertDialog,
  AlertDialogBody,
  AlertDialogContent,
  AlertDialogFooter,
  AlertDialogHeader,
  AlertDialogOverlay,
  Box,
  Button,
  FormControl,
  FormHelperText,
  FormLabel,
  Input,
  Spinner,
  Text,
  VStack,
} from '@chakra-ui/react';
import { useForm } from 'react-hook-form';
import { TextareaField } from '../TextareaField';
import { ErrorMessage } from '@hookform/error-message';
import { ErrorDisplay } from '../ErrorDisplay';

interface ModalBoardProps {
  isOpen: boolean;
  onClose: () => void;
  onConfirm: (data?: string) => void;
  showReason: boolean;
  showDynamicForm?: boolean;
  dynamicForm?: string;
  isDisabled?: boolean;
  isLoading?: boolean;
  setReason: (data: string) => void;
}

interface IDynamicFormProps {
  [key: string]: string;
}

const ModalBoard = (props: ModalBoardProps): JSX.Element => {
  const cancelRef = useRef(null);
  const {
    isDisabled = false,
    isLoading = false,
    isOpen,
    onClose,
    onConfirm,
    showReason = false,
    showDynamicForm = false,
    dynamicForm = '',
    setReason,
  } = props;

  const {
    register,
    handleSubmit,
    reset,
    formState: { errors },
  } = useForm({
    criteriaMode: 'all',
  });

  useEffect(() => {
    reset();
  }, [reset, showDynamicForm]);

  const dynamicFormParse = useMemo(() => {
    if (showDynamicForm) {
      return JSON.parse(dynamicForm);
    }
    return {};
  }, [dynamicForm, showDynamicForm]);

  const toDisplayName = (inputName: string) => {
    return inputName.replace(/([a-z])([A-Z])/g, '$1 $2');
  };

  const renderFormContent = (data: IDynamicFormProps[] | undefined) => {
    return data?.map(function (element, ind) {
      return (
        <FormControl key={ind}>
          <FormLabel fontSize={16} my={1} fontWeight="normal">
            {toDisplayName(element.name)}
            {element.isRequired ? (
              <FormHelperText my={1} style={{ color: 'red' }} as="span">
                {' '}
                *
              </FormHelperText>
            ) : (
              ''
            )}
          </FormLabel>
          <TextareaField
            {...register(element.name, {
              required: element.isRequired
                ? `${toDisplayName(element.name)} is Required`
                : false,
            })}
          />
          <ErrorMessage
            errors={errors}
            name={element.name}
            render={({ message }) => <ErrorDisplay message={message} />}
          />
        </FormControl>
      );
    });
  };

  const onSubmit = async (data: IDynamicFormProps) => {
    if (showDynamicForm) {
      for (const item of dynamicFormParse) {
        if (Object.prototype.hasOwnProperty.call(data, item.name)) {
          item.data = data[item.name];
        }
      }
      onConfirm(JSON.stringify(dynamicFormParse));
      return;
    }

    onConfirm();
  };

  return (
    <>
      <AlertDialog
        isOpen={isOpen}
        leastDestructiveRef={cancelRef}
        onClose={onClose}
      >
        <AlertDialogOverlay>
          <AlertDialogContent>
            <AlertDialogHeader fontSize="lg" fontWeight="bold" pb={1}>
              Update status
            </AlertDialogHeader>

            <AlertDialogBody>
              <Text
                whiteSpace="nowrap"
                fontSize="md"
                fontWeight="medium"
                pb={0}
              >
                Do you want to update status ?
              </Text>
              {showDynamicForm && (
                <form
                  style={{ width: '100%', margin: '10px 0' }}
                  onSubmit={handleSubmit(onSubmit)}
                >
                  <VStack spacing="14px" alignItems="flex-start">
                    {renderFormContent(dynamicFormParse)}
                  </VStack>
                </form>
              )}
              {showReason && (
                <Box mt={5}>
                  <Text
                    whiteSpace="nowrap"
                    fontSize="md"
                    fontWeight="medium"
                    pb={2}
                    pt={0}
                  >
                    Reason *
                  </Text>
                  <Input
                    type="text"
                    onChange={(e) => setReason(e.target.value)}
                  />
                </Box>
              )}
            </AlertDialogBody>

            <AlertDialogFooter>
              <Button onClick={onClose}>Cancel</Button>
              <Button
                colorScheme="red"
                onClick={handleSubmit(onSubmit)}
                ml={3}
                isDisabled={isDisabled || isLoading}
              >
                {isLoading ? <Spinner /> : 'Confirm'}
              </Button>
            </AlertDialogFooter>
          </AlertDialogContent>
        </AlertDialogOverlay>
      </AlertDialog>
    </>
  );
};

export default ModalBoard;
