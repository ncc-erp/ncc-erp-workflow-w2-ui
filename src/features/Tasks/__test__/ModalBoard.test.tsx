import { render } from '@testing-library/react';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import ModalBoard from 'common/components/Boards/ModalBoard';

// jest.mock('../../../api/apiHooks/index', () => ({
//     useAxios: jest.fn(),
// }));

// jest.mock('../../../api/axiosInstant', () => ({
//     VITE_API_BASE_URL: '/api',
// }));

// jest.mock('api/apiHooks/taskHooks', () => ({

// }))

test('Request My Requests Page', () => {
  const queryClient: QueryClient = new QueryClient();
  const { baseElement } = render(
    <QueryClientProvider client={queryClient}>
      <ModalBoard
        isOpen={true}
        onClose={() => jest.fn()}
        onConfirm={() => jest.fn()}
        showReason={true}
        setReason={() => jest.fn()}
        showDynamicForm={false}
        dynamicForm=""
        isDisabled={false}
        isLoading={false}
      />
    </QueryClientProvider>
  );
  expect(baseElement).toMatchSnapshot();
});
