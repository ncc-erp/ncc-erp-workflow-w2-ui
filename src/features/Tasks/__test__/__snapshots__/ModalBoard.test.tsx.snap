// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Request My Requests Page 1`] = `
<body>
  <div
    aria-hidden="true"
    data-aria-hidden="true"
  />
  <div
    class="chakra-portal"
  >
    <div
      class="chakra-modal__overlay css-13d1tkw"
      style="opacity: 0;"
    >
      <div
        aria-hidden="true"
        data-aria-hidden="true"
        data-focus-guard="true"
        style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
        tabindex="0"
      />
      <div
        data-focus-lock-disabled="false"
      >
        <div
          class="chakra-modal__content-container css-1oxhx2p"
          tabindex="-1"
        >
          <section
            aria-describedby="chakra-modal--body-:r0:"
            aria-labelledby="chakra-modal--header-:r0:"
            aria-modal="true"
            class="chakra-modal__content css-6yiygt"
            id="chakra-modal-:r0:"
            role="alertdialog"
            style="opacity: 0; transform: scale(0.95) translateZ(0);"
            tabindex="-1"
          >
            <header
              class="chakra-modal__header css-1rwr1rh"
              id="chakra-modal--header-:r0:"
            >
              Update status
            </header>
            <div
              class="chakra-modal__body css-0"
              id="chakra-modal--body-:r0:"
            >
              <p
                class="chakra-text css-1hqtvl7"
              >
                Do you want to update status ?
              </p>
              <div
                class="css-0"
                style="padding: 4px 0px;"
              />
              <div
                class="css-odz94x"
              >
                <p
                  class="chakra-text css-wc5dkf"
                >
                  Reason
                  <span
                    style="color: red;"
                  >
                     *
                  </span>
                </p>
              </div>
            </div>
            <footer
              class="chakra-modal__footer css-192qrng"
            >
              <button
                class="chakra-button css-4xx2wk"
                type="button"
              >
                Cancel
              </button>
              <button
                class="chakra-button css-1i56vmk"
                type="button"
              >
                Confirm
              </button>
            </footer>
          </section>
        </div>
      </div>
      <div
        aria-hidden="true"
        data-aria-hidden="true"
        data-focus-guard="true"
        style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
        tabindex="0"
      />
    </div>
  </div>
</body>
`;
